### ISSUE TITLE
add benchmarks

### ISSUE THREAD

**Issue by ev-br** [ev-br/sparr#17]
and actually compare to, say, `dok_matrix`.

**Comment by pv** [comment]
Probably best to compare against [pysparse](pysparse.sourceforge.net).ll_mat

**Comment by pv** [comment]
Also, memory usage would be important to understand and document

**Comment by ev-br** [comment]
Very crude for now, just %timeit and %memit, using the example from pysparse docs:
https://github.com/ev-br/sparr/blob/master/benchmarks.ipynb

Basically, this is now about 2x slower than `ll_mat` for single-element `__setitem__` + python looping overhead. The memory consumption is not great, but not terrible either, at least on this problem size.

**Comment by pv** [comment]
Also see https://github.com/scipy/scipy/pull/6004

**Comment by pv** [comment]
The memory usage may be a non-trivial disadvantage for std::map

**Comment by ev-br** [comment]
It could be, yes. Like I was saying over at scipy issue, I wonder what's the deal with the Wikipedia link map: it's being invoked for both sides of the argument.

**Comment by ev-br** [comment]
On a more serious note, there's no question that a sorted vector of sorted vectors wins *if* there's an idea of the number of rows/columns. Otherwise it's either reallocation or back to essentially the same memory as `std::map`, no?

**Comment by maciejkula** [comment]
Vec of vecs has a per-row overhead, whereas `std::map` has a per-entry overhead. This matters if you have a lot of entries.

**Comment by ev-br** [comment]
But you have to reallocate vectors if there's an insertion into a row which was previously empty and there were non-empty rows from both sides of it?

**Comment by maciejkula** [comment]
In https://github.com/scipy/scipy/pull/6004 I pre-allocate all the row vectors, so I pay the per-vector overhead up-front.

**Comment by ev-br** [comment]
Makes sense --- so you require an a priori estimate of the number of rows?

**Comment by maciejkula** [comment]
I just realised you've been working on this recently, apologies for stepping on your toes! I scanned scipy PRs so as to not duplicate work, but yours was separate :(

**Comment by maciejkula** [comment]
Yes. This can be relaxed with some work.

**Comment by ev-br** [comment]
Re: toes: No problem whatsoever! Hope you don't feel I'm stepping on yours.

**Comment by ev-br** [comment]
In your solution with preallocation: you can probably just have a second vector with row indices. The overhead is at most `nrows`, but there's no reallocation at all.

**Comment by maciejkula** [comment]
Yep, or have a `resize` call that allocates new vectors when necessary.

**Comment by ev-br** [comment]
https://github.com/ev-br/sparr/pull/40#issuecomment-208940574 for some numbers on constructing a Possoin 2D matrix from `pysparse` docs. 

I've no idea why memory benchmarks report all zeros. From `peakmem`, fast_lil_mat actually looks good :-). Timewise, `ll_mat` wins hands down.
